ROS2 module for aztarna

olaldiko opened the pull request.

LanderU commented:
This is unfinished work, right?

LanderU commented:
Crystal error:

```bash
root@wks4:~/aztarna# aztarna -t ROS2
aztarna.cmd - Exception occurred during execution
Traceback (most recent call last):
  File "/usr/local/bin/aztarna", line 11, in <module>
    load_entry_point('aztarna==1.0', 'console_scripts', 'aztarna')()
  File "/usr/local/lib/python3.7/dist-packages/aztarna-1.0-py3.7.egg/aztarna/cmd.py", line 90, in main
    self.stdin = sys.stdin
  File "/usr/local/lib/python3.7/dist-packages/aztarna-1.0-py3.7.egg/aztarna/cmd.py", line 82, in main
    is done automatically. The optional arguments stdin and stdout
  File "/usr/local/lib/python3.7/dist-packages/aztarna-1.0-py3.7.egg/aztarna/ros/ros2/scanner.py", line 100, in scan
  File "/opt/ros/crystal/lib/python3.6/site-packages/rclpy/__init__.py", line 25, in init
    context = get_default_context() if context is None else context
  File "/opt/ros/crystal/lib/python3.6/site-packages/rclpy/utilities.py", line 31, in get_default_context
    g_default_context = Context()
  File "/opt/ros/crystal/lib/python3.6/site-packages/rclpy/context.py", line 28, in __init__
    from rclpy.impl.implementation_singleton import rclpy_implementation
  File "/opt/ros/crystal/lib/python3.6/site-packages/rclpy/impl/implementation_singleton.py", line 31, in <module>
    rclpy_implementation = _import('._rclpy')
  File "/opt/ros/crystal/lib/python3.6/site-packages/rclpy/impl/__init__.py", line 21, in _import
    return importlib.import_module(name, package='rclpy')
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ModuleNotFoundError: No module named 'rclpy._rclpy
```

Dashing error:

```bash
root@wks4:~/aztarna# aztarna -t ROS2
aztarna.cmd - Exception occurred during execution
Traceback (most recent call last):
  File "/usr/local/bin/aztarna", line 11, in <module>
    load_entry_point('aztarna==1.0', 'console_scripts', 'aztarna')()
  File "/usr/local/lib/python3.7/dist-packages/aztarna-1.0-py3.7.egg/aztarna/cmd.py", line 90, in main
    self.stdin = sys.stdin
  File "/usr/local/lib/python3.7/dist-packages/aztarna-1.0-py3.7.egg/aztarna/cmd.py", line 82, in main
    is done automatically. The optional arguments stdin and stdout
  File "/usr/local/lib/python3.7/dist-packages/aztarna-1.0-py3.7.egg/aztarna/ros/ros2/scanner.py", line 100, in scan
  File "/opt/ros/dashing/lib/python3.6/site-packages/rclpy/__init__.py", line 70, in init
    context = get_default_context() if context is None else context
  File "/opt/ros/dashing/lib/python3.6/site-packages/rclpy/utilities.py", line 31, in get_default_context
    g_default_context = Context()
  File "/opt/ros/dashing/lib/python3.6/site-packages/rclpy/context.py", line 28, in __init__
    from rclpy.impl.implementation_singleton import rclpy_implementation
  File "/opt/ros/dashing/lib/python3.6/site-packages/rclpy/impl/implementation_singleton.py", line 31, in <module>
    rclpy_implementation = _import('._rclpy')
  File "/opt/ros/dashing/lib/python3.6/site-packages/rclpy/impl/__init__.py", line 21, in _import
    return importlib.import_module(name, package='rclpy')
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
ModuleNotFoundError: No module named 'rclpy._rclpy'
```
```

LanderU commented:
I think the problem are the python versions.

olaldiko commented:
@LanderU, I have isolated the dependency on Python3.7, so now it should be installable on default ROS2 python installations. I have also limited the maximum ROS domain ID to 232, it was throwing an error on @borkenerice's machine. I have read that this is the limit for several DDS implementations (FastRTPS and RTI), is that correct? There's not much documentation about it.

LanderU commented:
You're right, https://github.com/ros2/rmw_fastrtps/issues/261

LanderU commented on aztarna/ros/ros2/scanner.py:
```diff
@@ -0,0 +1,170 @@
+            raise Exception('ROS2 needs to be installed and sourced to run ROS2 scans')
+
+        for i in range(0, max_ros_domain_id):
+            os.environ['ROS_DOMAIN_ID'] = str(i)
```
What about implement the same for different RMW implementation?

olaldiko replied:
Shouldn't they be interoperable between different middlewares? At least to some basic level

LanderU replied:
Not in all ROS2 versions AFAIK.

olaldiko replied:
We are attempting to create a method to detect the installed RMW implementations, but without success. The first idea was to alter the RMW_IMPLEMENTATION environment variable prior to calling rclpy.init, in a try-except clause, but rclpy finishes the process on error.
Any idea on how we could implement it?? Listing installed packages maybe?

LanderU replied:
Let me have a look.

olaldiko replied:
@LanderU , I have created a preliminary version of the scanner, but I'm getting opensplice and connext related errors when running it. Let's see if we can find the cause.
![image](https://user-images.githubusercontent.com/1711272/59699919-ae116700-91f2-11e9-9422-05f99e48be6d.png)

LanderU replied:
Humm, I'll have a look into this this night.

LanderU replied:
```bash
[+] Host found in Domain ID 40
	Topics:
	Nodes:
		Node Name: _ros2cli_daemon_40 	|	 Namespace: /
		Node Name: talker 	|	 Namespace: /
--------------------------------------------------------------------------------
[+] Host found in Domain ID 40
	Topics:
	Nodes:
		Node Name: _ros2cli_daemon_40 	|	 Namespace: /
		Node Name: talker 	|	 Namespace: /
--------------------------------------------------------------------------------
[+] Host found in Domain ID 40
	Topics:
	Nodes:
		Node Name: _ros2cli_daemon_40 	|	 Namespace: /
		Node Name: talker 	|	 Namespace: /
--------------------------------------------------------------------------------
```

Using OpenSplice.

LanderU replied:
I'm going to try to create a patch.

LanderU replied:
```bash
========================================================================================
Context     : DDS::DomainParticipantFactory::create_participant
Date        : 2019-06-18T19:50:19+0000
Node        : 3e4a113cc568
Process     : python3.6 <9951>
Thread      : main thread 7fd60e17a740
Internals   : DomainParticipantFactory.cpp/324/6.9.190403OSS///-1
----------------------------------------------------------------------------------------
Report      : Unable to connect to domain id = 1.
              The most common causes of this error are an incorrect configuration file or
              that OpenSpliceDDS is not running (when using shared memory mode).
Internals   : u_participantNew/u_participant.c/234/773/1560887419.960899293
----------------------------------------------------------------------------------------
Report      : Failed to start Spliced for domain 'ospl_sp_ddsi' within 0 seconds, result = U_RESULT_PRECONDITION_NOT_MET

Internals   : user::u_domain::startSpliceThread/u_domain.c/1253/781/1560887419.960924533

>>> [rcutils|error_handling.c:106] rcutils_set_error_state()
This error state is being overwritten:

  'failed to create domain participant, at /tmp/binarydeb/ros-dashing-rmw-opensplice-cpp-0.7.1/src/rmw_node.cpp:162, at /tmp/binarydeb/ros-dashing-rcl-0.7.4/src/rcl/node.c:326'

with this new error message:

  'rcl node's rmw handle is invalid, at /tmp/binarydeb/ros-dashing-rcl-0.7.4/src/rcl/node.c:464'

rcutils_reset_error() should be called after error handling to avoid this.
<<<
[ERROR] [rcl]: Failed to fini publisher for node: 1
```

LanderU replied:
https://istkb.adlinktech.com/article/configuring-shared-memory/

LanderU commented on aztarna/ros/ros2/scanner.py:
```diff
@@ -5,6 +5,7 @@
 
 default_topics = ['/rosout', '/parameter_events']
 max_ros_domain_id = 231
+rmw_implementations = ['rmw_opensplice_cpp', 'rmw_fastrtps_cpp', 'rmw_connext_cpp']
```
Take into account, there are other rmw implementation such as:https://github.com/ros2/rmw_dps, https://github.com/ros2/rmw_coredds

olaldiko commented:
Given the unstability of using multiple rmw implementations I have moved the multi-rmw functionality to a separate branch. We will tackle the usage of multiple rmw implementations after base scanner merge. I have added functionality for getting service information from hosts.

LanderU commented:
So, is this finished? If so, I'll test it this night and if we're okay, merge it.

olaldiko commented:
I think so! Let's hope that everything is ok and we will be able to merge it tomorrow.

LanderU commented on Dockerfile:
```diff
@@ -17,7 +17,7 @@
 ####
 
-FROM ubuntu:18.04
+FROM osrf/ros2:nightly
```
I think is better if we use, ros:dashing for now

LanderU commented on Dockerfile:
```diff
@@ -34,6 +34,8 @@ RUN rm -rf /var/lib/apt/lists/*
 # RUN cd /root/aztarna && git checkout ${AZTARNA_COMMIT} && python3.7 setup.py install
-RUN cd /root/aztarna && python3.7 setup.py install
+RUN cd /root/aztarna && python3 setup.py install
+RUN echo "source /opt/ros/dashing/setup.bash" >> /root/.bashrc
```
Can I rework the docker part?

LanderU replied:
I've pushed some changes related to the docker, let me know If you're happy with them.

LanderU replied:
Docker output wit the latest changes:

```bash
lander@wks4:/tmp/aztarna(ros2_scanner)$ docker run  -it --rm --net=host aztarna -t ros2
[+] Host found in Domain ID 0
	Topics:
	Services:
		Service Name: /aztarna/describe_parameters 	|	 Service Type: ['rcl_interfaces/srv/DescribeParameters']
		Service Name: /aztarna/get_parameter_types 	|	 Service Type: ['rcl_interfaces/srv/GetParameterTypes']
		Service Name: /aztarna/get_parameters 	|	 Service Type: ['rcl_interfaces/srv/GetParameters']
		Service Name: /aztarna/list_parameters 	|	 Service Type: ['rcl_interfaces/srv/ListParameters']
		Service Name: /aztarna/set_parameters 	|	 Service Type: ['rcl_interfaces/srv/SetParameters']
		Service Name: /aztarna/set_parameters_atomically 	|	 Service Type: ['rcl_interfaces/srv/SetParametersAtomically']
	Nodes:
		Node Name: _ros2cli_daemon_0 	|	 Namespace: /
		Node Name: robot_state_publisher 	|	 Namespace: /
		Node Name: hros_cognition_kaladin 	|	 Namespace: /
		Node Name: minimal_action_server 	|	 Namespace: /
--------------------------------------------------------------------------------
[+] Host found in Domain ID 55
	Topics:
	Services:
		Service Name: /aztarna/describe_parameters 	|	 Service Type: ['rcl_interfaces/srv/DescribeParameters']
		Service Name: /aztarna/get_parameter_types 	|	 Service Type: ['rcl_interfaces/srv/GetParameterTypes']
		Service Name: /aztarna/get_parameters 	|	 Service Type: ['rcl_interfaces/srv/GetParameters']
		Service Name: /aztarna/list_parameters 	|	 Service Type: ['rcl_interfaces/srv/ListParameters']
		Service Name: /aztarna/set_parameters 	|	 Service Type: ['rcl_interfaces/srv/SetParameters']
		Service Name: /aztarna/set_parameters_atomically 	|	 Service Type: ['rcl_interfaces/srv/SetParametersAtomically']
	Nodes:
		Node Name: talker 	|	 Namespace: /
--------------------------------------------------------------------------------
```

LanderU commented on aztarna/cmd.py:
```diff
@@ -47,6 +48,8 @@ def main():
                 scanner.initialize_shodan()
         elif args.type.upper() == 'ROSIN':
             scanner = ROSIndustrialScanner()
+        elif args.type == 'ROS2' or args.type == 'ros2':
```
We need to have a look on this:

```bash
root@676deddf42e8:~/aztarna# aztarna -t ros2
aztarna.cmd - Exception occurred during execution
Traceback (most recent call last):
  File "/usr/local/bin/aztarna", line 11, in <module>
    load_entry_point('aztarna==1.2', 'console_scripts', 'aztarna')()
  File "/usr/local/lib/python3.6/dist-packages/aztarna-1.2-py3.6.egg/aztarna/cmd.py", line 91, in main
    if stdout is not None:
  File "/usr/local/lib/python3.6/dist-packages/aztarna-1.2-py3.6.egg/aztarna/cmd.py", line 65, in main
    identchars = IDENTCHARS
  File "/usr/local/lib/python3.6/dist-packages/aztarna-1.2-py3.6.egg/aztarna/ros/ros2/scanner.py", line 19, in scan_pipe_main
NotImplementedError
```

LanderU replied:
Fixed via: https://github.com/aliasrobotics/aztarna/pull/26/commits/2c8a9749fe40af826dba5a7e87789a0dfe4b27ae

olaldiko commented:
I have updated the readme and the docs. I think that we are done with this! If so, merge the PR and I will update the PyPi and RTD versions!
